### PR TITLE
Clarifications & tests for compact trees

### PR DESCRIPTION
**Pull request by daviddrysdale**
- [ ] I have updated the [CHANGELOG](CHANGELOG.md).
  - Adjust the draft version number according to [semantic versioning](https://semver.org/) rules.
- [ ] I have updated [documentation](docs/) accordingly.

### CODE REVIEW

**Comment by daviddrysdale** [comment]
Prob'ly best reviewed commit-by-commit, and the final commit ("merkle/compact: drop unhittable arm") needs the closest review.

**Comment by Martin2112** [comment]
There's a couple of cases where error paths aren't tested but that might be tricky as errors aren't passed back from everything.

**Comment by Martin2112** [comment]
Your reasoning that it can't get to that point seems sound but I need to look at that last commit a bit more. 

If the 'bit > 0' test wasn't there it would be a duplicate call to the setter function as the comment implies.

**Comment by Martin2112** [comment]
Empirical testing with a version rigged to crash shows that branch not being taken up to 10M additions.

**Comment by Martin2112** [comment]
Might as well ding the TODO(al) for lazy evaluation. Seems unlikely to be done and not sure we can as we need the side effects of the function calls.

File: merkle/compact/tree.go [hunk 1]
```diff
@@ -37,30 +37,40 @@ func (r RootHashMismatchError) Error() string {
...
-// NewTreeWithState creates a new compact Tree for the passed in |size|.
+// NewTreeWithState creates a new compact Tree for the passed in size.
 // This can fail if the nodes required to recreate the tree state cannot be fetched or the calculated
-// root hash after population does not match the value we expect.
-// |f| will be called a number of times with the co-ordinates of internal MerkleTree nodes whose hash values are
-// required to initialize the internal state of the compact Tree.  |expectedRoot| is the known-good tree root
-// of the tree at |size|, and is used to verify the correct initial state of the compact Tree after initialisation.
-func NewTreeWithState(hasher hashers.LogHasher, size int64, f GetNodeFunc, expectedRoot []byte) (*Tree, error) {
+// root hash after population does not match the expected value.
+// getNodeFn will be called a number of times with the co-ordinates of internal MerkleTree nodes whose hash values are
```
Comment (pavelkalinnikov) [inline merkle/compact/tree.go]: coordinates?

File: merkle/compact/tree.go [hunk 2]
```diff
@@ -37,30 +37,40 @@ func (r RootHashMismatchError) Error() string {
...
-// Uses log(n) nodes to represent the current on-disk tree.
+// Uses log(size) nodes to represent the current on-disk tree.
 type Tree struct {
 	hasher hashers.LogHasher
 	root   []byte
-	// the list of "dangling" left-hand nodes, NOTE: index 0 is the leaf, not the root.
+	// The list of "dangling" left-hand nodes, where entry [0] is the leaf.
+	// So: nodes[0] is the hash of a subtree of size 1=1<<0, if included.
+	//     nodes[1] is the hash of a subtree of size 2=1<<1, if included.
+	//     nodes[2] is the hash of a subtree of size 4=1<<2, if included.
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply merkle/compact/tree.go]: Done (assuming you mean to add spaces around operator).

File: merkle/compact/tree.go [hunk 3]
```diff
@@ -37,30 +37,40 @@ func (r RootHashMismatchError) Error() string {
...
+	//     ....
+	// Nodes are included if the tree size includes that power of two.
+	// For example, a tree of size 21 is built from subtrees of sizes
+	// 16 + 4 + 1, so nodes[1]==nodes[3]==nil.
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply merkle/compact/tree.go]: Done (assuming you mean to add spaces around operator).

File: merkle/compact/tree.go [hunk 4]
```diff
@@ -37,30 +37,40 @@ func (r RootHashMismatchError) Error() string {
...
 // This can fail if the nodes required to recreate the tree state cannot be fetched or the calculated
-// root hash after population does not match the value we expect.
-// |f| will be called a number of times with the co-ordinates of internal MerkleTree nodes whose hash values are
-// required to initialize the internal state of the compact Tree.  |expectedRoot| is the known-good tree root
-// of the tree at |size|, and is used to verify the correct initial state of the compact Tree after initialisation.
-func NewTreeWithState(hasher hashers.LogHasher, size int64, f GetNodeFunc, expectedRoot []byte) (*Tree, error) {
+// root hash after population does not match the expected value.
+// getNodeFn will be called a number of times with the co-ordinates of internal MerkleTree nodes whose hash values are
+// required to initialize the internal state of the compact Tree.  The expectedRoot is the known-good tree root
+// of the tree at size, and is used to verify the correct initial state of the compact Tree after initialisation.
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply merkle/compact/tree.go]: Pre-existing but done.

File: merkle/compact/tree.go [hunk 5]
```diff
@@ -249,12 +262,13 @@ func (t *Tree) AddLeafHash(leafHash []byte, f setNodeFunc) (int64, error) {
...
 }
 
-// Size returns the current size of the tree, that is, the number of leaves ever added to the tree.
+// Size returns the current size of the tree.
 func (t *Tree) Size() int64 {
 	return t.size
 }
 
 // Hashes returns a copy of the set of node hashes that comprise the compact representation of the tree.
+// A tree whose size is a power of two has no internal node hashes, so returns nil.
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply merkle/compact/tree.go]: I've tweaked the comment slightly but I'm not going to change behaviour in this PR.

File: merkle/compact/tree.go (hunk 1 above)
Comment (daviddrysdale) [reply merkle/compact/tree.go]: Pre-existing but done.

File: merkle/compact/tree.go [hunk 6]
```diff
@@ -225,13 +225,6 @@ func (t *Tree) AddLeafHash(leafHash []byte, setNodeFn setNodeFunc) (int64, error
 		if mask&1 == 0 {
 			// Just store the running hash here; we're done.
 			t.nodes[bit] = hash
-			// Don't re-write the leaf hash node (we've done it above already)
-			if bit > 0 {
-				// Store the (non-leaf) hash node
-				if err := setNodeFn(bit, index, hash); err != nil {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply merkle/compact/tree.go]: So it's unhittable **and** wrong?

[the remark this answers is not in the document]
Comment (daviddrysdale) [reply merkle/compact/tree.go]: BTW, I tried to explain the unhittableness in the last commit message

[the remark this answers is not in the document]
Comment (Martin2112) [reply merkle/compact/tree.go]: I don't know where this came from. Perhaps @AlCutter  can remember?

File: merkle/compact/tree_test.go [hunk 7]
```diff
@@ -230,8 +252,26 @@ func TestRootHashForVariousTreeSizes(t *testing.T) {
...
-			t.Errorf("Test (treesize=%v) got root %v, want %v", test.size, b64e(got), b64e(want))
+		if gotRoot := tree.CurrentRoot(); !bytes.Equal(gotRoot, test.wantRoot) {
+			t.Errorf("Test (treesize=%v) got root %v, want %v", test.size, b64e(gotRoot), b64e(test.wantRoot))
+		}
+		if gotDepth := tree.Depth(); gotDepth != test.wantDepth {
+			t.Errorf("Test (treesize=%v).Depth()=%d, want %d", test.size, gotDepth, test.wantDepth)
+		}
+		t.Log(tree)
+		if isPerfectTree(test.size) {
+			// A perfect tree should have a single hash at the highest bit that is just
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply merkle/compact/tree_test.go]: Tweaked the comment.

File: merkle/compact/tree_test.go [hunk 8]
```diff
@@ -131,6 +131,9 @@ func TestLoadingTreeFailsBadRootHash(t *testing.T) {
 	if err == nil || !ok {
 		t.Errorf("Did not return correct error type on root mismatch: %v", err)
 	}
+	if !strings.Contains(err.Error(), "mismatch") {
+		t.Errorf("Error %v doesn't mention mismatch", err)
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply merkle/compact/tree_test.go]: Done.

File: merkle/compact/tree.go [hunk 9]
```diff
@@ -166,34 +175,38 @@ func (t *Tree) recalculateRoot(f setNodeFunc) error {
...
 	return seq, h, err
 }
 
-// AddLeafHash adds the specified |leafHash| to the tree.
-// |f| is a callback which will be called multiple times with the full MerkleTree coordinates of nodes whose hash should be updated.
-func (t *Tree) AddLeafHash(leafHash []byte, f setNodeFunc) (int64, error) {
+// AddLeafHash appends the specified Merkle leaf hash to the tree.
+// setNodeFn is a callback which will be called multiple times with the full MerkleTree coordinates
+// of nodes whose hash should be updated.
+func (t *Tree) AddLeafHash(leafHash []byte, setNodeFn setNodeFunc) (int64, error) {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply merkle/compact/tree.go]: Done.

**Comment by pavelkalinnikov** [comment]
David, are you planning to merge this? Sorry, I didn't realise it wasn't, and it looks like I introduced some conflicts.

**Comment by daviddrysdale** [comment]
I was planning to merge most of it, but I might wimp out of the final commit.  I'll rebase and resolve the conflicts at some point next week.

**Comment by daviddrysdale** [comment]
Rebased & dropped the final commit as it's due to get re-written anyway, PTAL.

File: merkle/compact/tree.go [hunk 10]
```diff
@@ -111,25 +122,41 @@ func NewTreeWithState(hasher hashers.LogHasher, size int64, f GetNodeFunc, expec
...
 	}
-	return &r
 }
 
 // CurrentRoot returns the current root hash.
 func (t *Tree) CurrentRoot() []byte {
 	return t.root
 }
 
+// String describes the internal state of the compact Tree.
```
Comment (Martin2112) [inline merkle/compact/tree.go]: Was this one of the things @pavelkalinnikov removed last week?

Comment (daviddrysdale) [reply merkle/compact/tree.go]: This commit used to be "convert DumpNodes to String" and now is just "add String()", but I added use of it in the tests so it's no longer unused.  Could just delete it, but a stringer method can be helpful.

Comment (Martin2112) [reply merkle/compact/tree.go]: OK.

**Pull request merged**